--- a/Source/WebKit/qt/Api/qwebsettings.cpp
+++ b/Source/WebKit/qt/Api/qwebsettings.cpp
@@ -57,13 +57,19 @@
 #include <wtf/FastMalloc.h>
 #include <wtf/text/WTFString.h>
 
+#if USE(SOUP)
+#include "ResourceHandle.h"
+#include <libsoup/soup.h>
+#endif
 
 
 QWEBKIT_EXPORT void qt_networkAccessAllowed(bool isAllowed)
 {
+#if !USE(SOUP)
 #ifndef QT_NO_BEARERMANAGEMENT
     WebCore::networkStateNotifier().setNetworkAccessAllowed(isAllowed);
 #endif
+#endif
 }
 
 class QWebSettingsPrivate {
@@ -1250,6 +1256,14 @@
 
     WebCore::makeAllDirectories(storagePath);
 
+#if USE(SOUP)
+    /* Store cookies in moz-compatible SQLite format */
+    SoupSession* session = WebCore::ResourceHandle::defaultSession();
+    SoupCookieJar* jar = soup_cookie_jar_db_new(WebCore::pathByAppendingComponent(storagePath, "cookies.sqlite").ascii().data(), FALSE);
+    soup_session_add_feature(session, SOUP_SESSION_FEATURE(jar));
+    g_object_unref(jar);
+#endif
+
     QWebSettings::setIconDatabasePath(storagePath);
     QWebSettings::setOfflineWebApplicationCachePath(storagePath);
     QWebSettings::setOfflineStoragePath(WebCore::pathByAppendingComponent(storagePath, "Databases"));
